## Syngonium Podophyllum Leaf Curvature
___

#### Prerequisite Papers:

[1] Bernd Schmidt, Plate theory for stressed heterogenous multilayers of finite bending energy, 2007, Journal de Mathématiques Pures et Appliquées, vol. 88, no. 1, pages 107-122, doi. https://doi.org/10.1016/j.matpur.2007.04.011


#### 1. Define Goal:
__ 19.10.2024 __
1. Calculate the (symmmetric) layer mismatch for the curvature measurements of Sygonium Podophyllum.
2. Derive a formula for the (symmetric) layer mismatch given measured radii.

__ 28.11.2024 __

3. Input the measured data concerning the number of cells in the lower and upper layers at different time stages
4. Calculate the Layer Mismatches from them
5. Calculate the predicted radius and compare to the measured radius.

# __19.10.2024__
#### 2. Define Constants to use for the following measurements

In [211]:
YOUNGS_MODULUS = 5.963
POISSON_RATIO = 0.3

#### 3. Reading in the Data

In [212]:
import pandas as pd
import numpy as np

df = pd.read_excel("pilea data 2.0/Spod_leaf-curvature_final-measurements.xlsx", sheet_name="Data", skiprows=1)
df

leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.524540            0.165667   
1                   0.143433          0.422961            0.204100   
2                   0.171333          0.299805            0.222700   
3                   0.131333          0.385787            0.182000   
4                   0.144667          0.421659            0.205667   
5                   0.106000          0.663522            0.176333   
6                   0.106333          0.514107            0.161000   
7                   0.142667          0.443925            0.206000   
8                   0.149333          0.361607            0.203333   
9                   0.136333          0.369193            0.186667   
10                  0.131333          0.411168            0.185333   

    leaf_width_mm  plate thickness h_mm  
0          60.448              0.002741  
1          73.546              0.002775  
2          41.638              0.005348  
3          46.339              0.003928  
4          62.311              0.003301  
5          60.348              0.002922  
6          53.834              0.002991  
7          43.277              0.004760  
8          64.786              0.003139  
9          81.568              0.002288  
10         45.794              0.004047

#### 4. Translating the middle surface

The middle surface $\tau$ in the table above is the ratio between the *thickness_upper* and *thickness_lower*. We want to translate it to a number $\tilde{\tau}$ such that $[-1/2, \tilde{\tau}]$ and $[\tilde{\tau}, 1/2]$ are the rescaled top and bottom layer respectively (i.e their ratio is identical). We therefore have to following connection
$$\frac{1/2 - \tilde{\tau}}{\tilde{\tau} + 1/2} = \tau \quad \Leftrightarrow \quad \frac{1}{2} \cdot \frac{1-\tau}{\tau + 1} = \tilde{\tau}$$
We will replace $\tau$ with $\tilde{\tau}$ and just call the updated values $\tau$ again.

In [213]:
df["middle surface τ"] = 0.5 * (1 - df["middle surface τ"]) / (df["middle surface τ"] + 1)
df

leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.155936            0.165667   
1                   0.143433          0.202760            0.204100   
2                   0.171333          0.269346            0.222700   
3                   0.131333          0.221612            0.182000   
4                   0.144667          0.203404            0.205667   
5                   0.106000          0.101134            0.176333   
6                   0.106333          0.160455            0.161000   
7                   0.142667          0.192557            0.206000   
8                   0.149333          0.234426            0.203333   
9                   0.136333          0.230357            0.186667   
10                  0.131333          0.208633            0.185333   

    leaf_width_mm  plate thickness h_mm  
0          60.448              0.002741  
1          73.546              0.002775  
2          41.638              0.005348  
3          46.339              0.003928  
4          62.311              0.003301  
5          60.348              0.002922  
6          53.834              0.002991  
7          43.277              0.004760  
8          64.786              0.003139  
9          81.568              0.002288  
10         45.794              0.004047

#### 5. Some Theory & A Simple Scheme to determine the mismatch

The calculation below is based on the imported script *tubes.py* which takes in the layer mismatch among other variables and calculates an expected radius (under the assumption that the theory applies here) i.e $F(b_{\text{top}}, b_{\text{bot}}, \ldots) = r_{\text{calculated}}$. In our case we have $r_{\text{calculated}}$ available and want to determine a pair of $b_{\text{top}}, b_{\text{bot}}$. Due to the nature of this problem, we will have one degree of freedom. Let's therefore first regard symmetric layer mismatches where $b_{\text{top}} = - b_{\text{bot}}$.

Looking at the equations, it is not hard to see that $\kappa$ depends only on $b_{\text{top}}$ and $b_{\text{bot}}$ through $f_0$ and this depenence is linear! Indeed, we have
$$\kappa = \frac{ e_1^\top \mathcal{M}_0 f_0}{ e_1^\top \mathcal{M}_0 e_1}$$
which is linear in $f_0$. Next we have
$$f_0 = \mathcal{M}_0^{-1} (\mathcal{M}_2 \mathcal{M}_1^{-1} b_1 - b_2)$$
By reference [1], we may w.l.o.g choose $b_1 = 0$, therefore $f_0$ is linear in $b_2$. Finally we have
$$b_2 = \left(\frac{1}{2} \tau^2 - \frac{1}{8}\right) \mathcal{M}_{\text{bot}} b_{\text{bot}} - \left( \frac{1}{8} - \frac{1}{2}\tau^2 \right) \mathcal{M}_{\text{top}} b_{\text{top}}$$
By the assumption of symmetric layer mismatch, we have $b_{\text{top}} = -b_{\text{bot}}$. Therefore, using this property above immediately shows the linearity of $\kappa$ in $b$, i.e.
$$\nabla_b \kappa = \begin{pmatrix}\frac{\partial \kappa}{\partial b_{\text{top}}} & \frac{\partial \kappa}{\partial b_{\text{top}}} \end{pmatrix}^\top \equiv \text{const.}$$ 
Also it is clear that increasing the layer mismatch will decrease the output radius of $F$.
Therefore we can implement a very rudimentary scheme that succesively tries layer mismatches and adjusts by a factor in the correct direction, if the resulting output radius $r_{\text{calculated}}$ is too far away from the measurement.

Using this calculated layer mismatch we also add the relative deviation to the DataFrame:

In [214]:
#import code
from tubes import optimal_radius

def _apply_find_layer_mismatch(df_row):
    b_current = 1
    while True:
        current = optimal_radius(tau=df_row["middle surface τ"],
                                 h=df_row["plate thickness h_mm"],
                                 total_thickness=df_row["total thickness_mm"],
                                 btop=b_current,
                                 bbot=-b_current,
                                 youngs_modulus=YOUNGS_MODULUS,
                                 poisson_ratio=POISSON_RATIO,
                                 measured_radius=df_row["r_mm"])
        difference = current["calculated_radius"] - df_row["r_mm"]
        if abs(difference) < 0.001:
            break
        elif difference > 0:
            b_current *= 1.1
        elif difference < 0:
            b_current *= 0.9

    return pd.Series([b_current, current["relative_deviation"]])

df_mismatch_from_radius = df.copy()
df_mismatch_from_radius[["sym_layer_mismatch", "radius: relative_deviation"]] = df_mismatch_from_radius.apply(_apply_find_layer_mismatch, axis=1)
df_mismatch_from_radius["sym_layer_mismatch (rescaled)"] = df_mismatch_from_radius["plate thickness h_mm"] * df_mismatch_from_radius["sym_layer_mismatch"]
df_mismatch_from_radius


leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.155936            0.165667   
1                   0.143433          0.202760            0.204100   
2                   0.171333          0.269346            0.222700   
3                   0.131333          0.221612            0.182000   
4                   0.144667          0.203404            0.205667   
5                   0.106000          0.101134            0.176333   
6                   0.106333          0.160455            0.161000   
7                   0.142667          0.192557            0.206000   
8                   0.149333          0.234426            0.203333   
9                   0.136333          0.230357            0.186667   
10                  0.131333          0.208633            0.185333   

    leaf_width_mm  plate thickness h_mm  sym_layer_mismatch  \
0          60.448              0.002741            9.088789   
1          73.546              0.002775           10.556994   
2          41.638              0.005348            9.990951   
3          46.339              0.003928            6.054750   
4          62.311              0.003301            8.504015   
5          60.348              0.002922            8.643347   
6          53.834              0.002991           10.200661   
7          43.277              0.004760            5.287426   
8          64.786              0.003139           10.250327   
9          81.568              0.002288            9.149798   
10         45.794              0.004047            4.938927   

    radius: relative_deviation  sym_layer_mismatch (rescaled)  
0                     0.000197                       0.024909  
1                    -0.000313                       0.029297  
2                    -0.000415                       0.053436  
3                    -0.000039                       0.023780  
4                    -0.000438                       0.028069  
5                    -0.000349                       0.025255  
6                    -0.000452                       0.030507  
7                     0.000183                       0.025168  
8                    -0.000062                       0.032171  
9                    -0

#### 6. Explanation of the Layer Mismatch

The symmetric layer mismatch determines the expansion/shrinkage of the top/bottom layer. In our case, if we have a **rescaled** symmetric layer mismatch of $b \in \mathbb{R}$, then this means that the top layer and bottom layer want to shrink and expand by a factor of
$$\frac{1}{1+b},\qquad \text{and} \qquad \frac{1}{1-b}$$
respectively, which induces the curvature.

#### 7. Obtaining a Formula for the Layer Mismatch

The simple scheme is a good starting point to get the right order of magnitude. However we are interested in an exact formula for the layer mismatch given a measured radius:

\begin{align*}
    \Leftrightarrow \qquad & r_{\text{calculated}} \overset{!}{=} r_{\text{measured}} \\
    \Leftrightarrow \qquad & \frac{d}{\kappa h} = r_{\text{measured}} \\
    \Leftrightarrow \qquad & \kappa = \frac{d}{h}\frac{1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & \frac{\beta}{\alpha} = \frac{d}{h}\frac{1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & \beta = \frac{d}{h}\frac{\alpha}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & e_1^\top \mathcal{M}_0 f_0 = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
\end{align*}



here $d$ denotes the total plate/leaf thickess. By [1], we may w.l.o.g assume $b_1 = 0$, such that

\begin{align*}
    \Leftrightarrow \qquad & e_1^\top \mathcal{M}_0 (-\mathcal{M}_0^{-1}b_2) = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & -e_1^\top b_2 = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & -e_1^\top \left( \left(\frac{1}{2} \tau^2 - \frac{1}{8}\right) M_{\text{bot}} \begin{pmatrix} -b_{\text{sym}} \\ -b_{\text{sym}} \\ 0 \end{pmatrix} + \left(\frac{1}{8} - \frac{1}{2} \tau^2\right) M_{\text{top}} \begin{pmatrix} b_{\text{sym}} \\ b_{\text{sym}} \\ 0 \end{pmatrix} \right) = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
\end{align*}

In order to simplify the formula, we will assume that both layers are isotropic with the same isotropy constants i.e. we have $M_{\text{bot}} = M_{\text{top}} =: M_{\text{iso}}$ such that

\begin{align*}
    \Leftrightarrow \qquad & -e_1^\top \left( \left(\frac{1}{2} \tau^2 - \frac{1}{8}\right) M_{\text{bot}} \begin{pmatrix} -b_{\text{sym}} \\ -b_{\text{sym}} \\ 0 \end{pmatrix} + \left(\frac{1}{8} - \frac{1}{2} \tau^2\right) M_{\text{top}} \begin{pmatrix} b_{\text{sym}} \\ b_{\text{sym}} \\ 0 \end{pmatrix} \right) = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & \left( \tau^2 - \frac{1}{4} \right) e_1^\top M_{\text{iso}} \begin{pmatrix} b_{\text{sym}} \\ b_{\text{sym}} \\ 0 \end{pmatrix} = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & b_{\text{sym}}\left( \tau^2 - \frac{1}{4} \right) e_1^\top M_{\text{iso}} (e_1 + e_2) = \frac{d}{h}\frac{e_1^\top \mathcal{M}_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & b_{\text{sym}} = \frac{d}{h} \frac{e_1^\top \mathcal{M}_0 e_1}{e_1^\top M_{\text{iso}} (e_1 + e_2)} \frac{1}{\tau^2 - \frac{1}{4}} \frac{1}{r_{\text{measured}}}
\end{align*}

Finally due to symmetry and with abuse of notation, we may flip the sign of the righthand-side to arrive at
\begin{equation} b_{\text{sym}} = \frac{d}{h} \frac{e_1^\top \mathcal{M}_0 e_1}{e_1^\top M_{\text{iso}} (e_1 + e_2)} \frac{1}{\frac{1}{4} - \tau^2}\frac{1}{r_{\text{measured}}}, \qquad \tau \in \left(-\frac{1}{2}, \frac{1}{2} \right), \qquad \mathcal{M}_0 = \mathcal{M}_0(M_{\text{iso}}, \tau) \in \mathbb{R}^{3 \times 3}, \qquad M_{\text{iso}} = M_{\text{iso}}(E, \nu) \in \mathbb{R}^{3 \times 3} \end{equation}
here $E$ denotes Young's Modulus and $\nu$ denotes Poisson's Ratio. Notice  $b_{\text{sym}}$ refers to the lower layer mismatch after the sign change.

Furthermore the matrix $\mathcal{M}_0$ is given by
$$\mathcal{M}_0 = \mathcal{M}_3 - \mathcal{M}_2 \mathcal{M}_1^{-1} \mathcal{M}_2$$
If we assume both the top and bottom layer having the same isotropy constants, these matricies in turn are given by
$$\mathcal{M}_i = \int_{-1/2}^{1/2} t^{i-1} M_{\text{iso}} dx \qquad i \in \{1,2,3\}$$
If we on the other hand assume both materials have different isotropy constants, above we integrate from $-1/2$ to $\tau$ over $M_{\text{bot}}$ and from $\tau$ to $1/2$ over $M_{\text{top}}$ where these matricies are derived the same as $M_{\text{iso}}$, but with the respective isotropy parameters of that specific layer (bot, top).

Finally $M_{\text{iso}}$ is just the matrix representation of the quadratic form $Q_2$ in the dimensionally reduced energy (again if we assume both layers share the same isotropy constants) which is exactly the same as in [1].

#### Intermezzo:

We generalize the formula if the material paramters are different. In particular we assume that the Poisson Ratio is identical, but the *Young's Modulus* is different.

It can be shown that $M_{\text{top}} = E_{\text{top}} M_{\text{iso}}^{(1)}$ where $M_{\text{iso}}^{(1)}$ is the isotropic material matrix when we choose the *Young's Modulus* as $E = 1$.

\begin{align*}
    \Leftrightarrow \qquad & -e_1^\top (M_{\text{top}} + M_{\text{bot}})\left(\left(\frac{1}{8} - \frac{1}{2} \tau^2\right) \begin{pmatrix} b_{\text{sym}} \\ b_{\text{sym}} \\ 0 \end{pmatrix} \right) = \frac{d}{h}\frac{e_1^\top M_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & \left( \tau^2 - \frac{1}{4} \right)\left(\frac{E_{\text{top}} + E_{\text{bot}}}{2}\right) e_1^\top M_{\text{iso}}^{(1)} \begin{pmatrix} b_{\text{sym}} \\ b_{\text{sym}} \\ 0 \end{pmatrix} = \frac{d}{h}\frac{e_1^\top M_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & b_{\text{sym}}\left( \tau^2 - \frac{1}{4} \right) \left(\frac{E_{\text{top}} + E_{\text{bot}}}{2}\right) e_1^\top M_{\text{iso}}^{(1)} (e_1 + e_2) = \frac{d}{h}\frac{e_1^\top M_0 e_1}{r_{\text{measured}}} \\
    \Leftrightarrow \qquad & b_{\text{sym}} = \frac{d}{h} \frac{e_1^\top M_0 e_1}{e_1^\top M_{\text{iso}}^{(1)} (e_1 + e_2)} \frac{2}{E_{\text{top}} + E_{\text{bot}}} \frac{1}{\tau^2 - \frac{1}{4}} \frac{1}{r_{\text{measured}}}
\end{align*}
___

This formula immediately tells us that the layer mismatch is inversely propotional to the measured radius. Also the term $\frac{1}{4} - \tau^2$ is interesting as it tells us that a non-even thickness distribution of the layers has to be compensated by the layer mismatch in order to arrive at he the same curvature i.e for the layer mismatch $b_{\text{sym}}$ to be as as small as possible, it is ideal if $\tau = 0$.

Let's test/verify this formula on the DataFrame above:

In [215]:
from numpy import abs as npabs
from tubes import calculate_Q_2_isotropic, calculate_Q_2_bar

## Retrieve the M matricies
## Notice that we choose them to be identical in accordance with the assumption above.
M_bot, M_top = calculate_Q_2_isotropic(youngs_modulus=YOUNGS_MODULUS, poisson_ratio=POISSON_RATIO)

def _apply_calculate_mismatch_using_formula(df_row, M_bot, M_top):
    assert (M_bot == M_top).all()
    M0, _, _, _ = calculate_Q_2_bar(tau=df_row["middle surface τ"], M_bot=M_bot, M_top=M_top)
    return (df_row["total thickness_mm"] / df_row["plate thickness h_mm"] * M0[0, 0]
            / (M_bot[0, 0] + M_bot[1, 0]) / (0.25 - df_row["middle surface τ"] ** 2) / df_row["r_mm"])

df_mismatch_from_radius_validate = pd.DataFrame({
    "mismatch by formula": df_mismatch_from_radius.apply(_apply_calculate_mismatch_using_formula, axis=1, args=(M_bot, M_top)),
    "mismatch by simple scheme": df_mismatch_from_radius["sym_layer_mismatch"]})
df_mismatch_from_radius_validate["Difference"] = npabs(df_mismatch_from_radius_validate["mismatch by formula"] - df_mismatch_from_radius_validate["mismatch by simple scheme"])
df_mismatch_from_radius_validate



mismatch by formula  mismatch by simple scheme  Difference
0              9.090579                   9.088789    0.001790
1             10.553695                  10.556994    0.003299
2              9.986802                   9.990951    0.004149
3              6.054511                   6.054750    0.000239
4              8.500291                   8.504015    0.003724
5              8.640328                   8.643347    0.003019
6             10.196052                  10.200661    0.004608
7              5.288395                   5.287426    0.000969
8             10.249695                  10.250327    0.000632
9              9.147600                   9.149798    0.002197
10             4.938550                   4.938927    0.000377

We see that our formula indeed captures the layer mismatch. The difference between the two columns is explained by the fact the the mismatch using the simple scheme is not exact as it terminates once $|r_{measured} - r_{output}| < \varepsilon$, thus we expect $|b_{formula} - b_{output}| < \delta$ for a $\delta > 0$. More specifically it is a continouity-$\delta = \delta(\varepsilon, F)$. Here $b_{output}$ is the output of the simple scheme above and $b_{formula}$ is the output of equation (1).

Finally we export the DataFrame:

In [216]:
df_mismatch_from_radius.to_excel("results_19_10_2024_nick.xlsx", index=False)

# __28.11.2024__ 

The goal of the next section is to use the measured data on the cell numbers and cell dimension to calculate a layer mismatch. 
This layer mismatch shall then be used to calculate a radius according to the model above.

The measured data is not complete. For instance we do not have the cell dimensions for leaf stage A and sample C5. In order to
calculate such a layer mismatch for a specific sample we need:

- cell number at stages A and (J0 or J1) denoted as $n_{A,\text{bot}}$, $n_{A,\text{top}}$ and $n_{J,\text{bot}}$, $n_{J,\text{top}}$
- cell width at stages A and (J0 or J1) denoted as $d_{A,\text{bot}}$, $d_{A,\text{top}}$ and $d_{J, \text{bot}}$, $d_{J, \text{top}}$

The layer mismatch here is not atomistic but rather microscopic, nevertheless using these values we want to determine a "length scale" $\epsilon$ 
for both layers and both stages A and (J0 or J1) and later consider their ratio. There are different approaches to choosing a length scale. 
The one we will regard here is

$$\epsilon_{A,\text{bot}} = d_{A,\text{bot}} \cdot n_{A,\text{bot}}, \qquad \epsilon_{A,\text{top}} = d_{A,\text{top}} \cdot n_{A,\text{top}}, \qquad \epsilon_{J,\text{bot}} = d_{J,\text{bot}} \cdot n_{J,\text{bot}}, \qquad \epsilon_{J,\text{top}} = d_{J,\text{top}} \cdot n_{J,\text{top}}$$

In words, we assume that the curvature is induced by both new cells spawning in both layers as well as existing cells changing in size, causing an internal strain where the layers meet.

In the Nanotube case, the curvature is created by the layer mismatch once the layers are released from the substrate. The mismatch is known by knowing the distance of the atoms of the material used. 
In this setting we need a different approach. That is, we choose to compare the microscopic stucture sizes in order to deduce a layer mismatch for the model. In particular,
we have to time stages, the rolled up leaf time stage and the rolled out leaf time stage. The latter will be the base stage in which we will assume that there is no internal stress i.e no layer mismatch. 
The rolled up leaf time stage will then be regarded relative to the base stage in the model.

Using the length scales above, we will calculate the ratio compared to the base time stage:

$$\phi_{\text{bot}} = \frac{\epsilon_{J,\text{bot}}}{\epsilon_{A,\text{bot}}}, \qquad \phi_{\text{top}} = \frac{\epsilon_{J,\text{top}}}{\epsilon_{A,\text{top}}}$$

### 8. Reading the mismatch data

In [217]:
df_cell_number_data = pd.read_excel("pilea data 3.0/Spod_total-cellnumb_per-leaf-stage_final.xlsx", sheet_name="Messungen")
df_cell_dimension_data = pd.read_excel("pilea data 3.0/Spod_leaf-cell-dimensions.xlsx", sheet_name="Measurements")
df_cell_dimension_data = df_cell_dimension_data.groupby(["leaf stage", "sample"]).agg({
    "cellwidth_upper_um": "mean",
    "cellthickn_upper_um": "mean",
    "cellwidth_lower_um": "mean",
    "cellthickn_lower_um": "mean"
}).reset_index()

df_cell_number_data.drop("counted on whole leaf or 1 leaf half?", axis=1, inplace=True)
df_mismatch_data = df_cell_number_data.merge(df_cell_dimension_data, on=["leaf stage", "sample"])
df_mismatch_data

leaf stage sample                                   full sample name  \
0         J-1    C01  2024-02-02_Spod_J-1_C01_PL6_Schnitt2_Bild2_5µ_...   
1         J-1    C02            SP J-1 CO2 PL2 Schnitt1 4x Maßstab500µm   
2         J-1    C04      SP J-1 CO4 PL1 Schnitt3 Bild1 4x Maßstab500µm   
3         J-1    C05  SP J-1 CO5 PL2 Schnitt3 Bild1 4x zus Maßstab10...   
4         J-1    C06  SP J-1 CO6 Pl 1 Schnitt1 Bild1 10x zus Maßstab...   
5          J0    C01  2024-01-22-SPod_J0 C01 PL12 Schnitt1 2x zus scale   
6          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_5µ_Aussch...   
7          J0    C03         SPJO CO3 PL1 Schnitt2 2x zus Maßstab2000µm   
8          J0    C04  SP JO CO4 PL4 Schnitt1 Bild1 Ende innnen 4x zu...   
9          J0    C06        SPJO CO6 PL1 Schnitt3 2x zus Maßstab 2000µm   
10         J1    C02  2024-03-01_SP-J1_CO2_PL21_Schnitt2_Bild2_Anfan...   
11         J1    C03          SP J1 CO3 PL5 Schnitt 2 2x Maßstab 2000µm   
12         J1    C04  SP J1 CO4 PL4 Schnitt2 Rechts Ende 4x zus Maßs...   
13         J1    C05  SPJ1CO5 PL2 Schnitt2 Bild Mitte  4x zus Maßsta...   
14         J1    C07        SP J1 CO7 PL2 Schnitt2 4x zus Maßstab2000µm   
15          A    C01                                 Cryo_Spod01_A_....   
16          A    C02                                 Cryo_Spod02_A_....   
17          A    C03                                 Cryo_Spod03_A_....   
18          A    C04                                 Cryo_Spod04_A_....   

    cellnumb_upper  cellnumb_lower  cellwidth_upper_um  cellthickn_upper_um  \
0             1603            1962             15.3347              19.3101   
1             1766            2306              9.3930              17.2804   
2             1003            1290              8.3457              13.8516   
3             1930            2381             10.3261              15.4045   
4             1060            1588              8.4541              12.2933   
5             2354            2850             18.7798              21.7363   
6             3548            3740             21.5511              25.6877   
7             1623            1888             21.0253              24.0297   
8             1913            2296             24.7538              17.9202   
9             2521            3068             21.1672              24.9619   
10            2076            2626             24.3624              22.4989   
11            1398            1612             26.1730              28.3114   
12            2714            3202             38.9263              46.9939   
13            3278            3984             28.4371              29.4316   
14            1644            2016             26.0582              25.9235   
15            2040            2025             32.4878              29.2753   
16            3498            3258             34.8680              36.0531   
17            2209            2160             40.1085              36.5838   
18            2543            2503             41.3778              34.3828   

    cellwidth_lower_um  cellthickn_lower_um  
0              11.8393              13.8604  
1              10.2669              13.2755  
2               9.6687              14.1294  
3               8.9633              14.2058  
4               6.0286              11.1746  
5              22.6897              14.5990  
6              21.1967              22.6050  
7              27.2222              18.1865  
8              29.0499              24.5890  
9              23.8196              24.0070  
10             31.0524              17.3958  
11             33.0915              24.1774  
12             51.4269              39.9112  
13             27.0775              23.4539  
14             22.3613              13.2372  
15             28.7986              26.9801  
16             35.5500              28.2512  
17             30.2746              22.5990  
18             26.1539              18.2120

#### 9. Making the mismatch data usable for the bilayer bending model

In [218]:
def _apply_calculate_mismatch_ratios(df_sample: pd.DataFrame):
    sample = df_sample.name
    j0_series = None
    j1_series = None
    if "A" in df_sample["leaf stage"].values:
        if "J0" in df_sample["leaf stage"].values:
            # gridsize refers to the variable epsilon above
            grid_size_top_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellnumb_upper"].values[0] * df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_upper_um"].values[0]
            grid_size_top_t1 = df_sample.loc[df_sample["leaf stage"] == "J0", "cellnumb_upper"].values[0] * df_sample.loc[df_sample["leaf stage"] == "J0", "cellwidth_upper_um"].values[0]
            grid_size_bot_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellnumb_lower"].values[0] * df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_lower_um"].values[0]
            grid_size_bot_t1 = df_sample.loc[df_sample["leaf stage"] == "J0", "cellnumb_lower"].values[0] * df_sample.loc[df_sample["leaf stage"] == "J0", "cellwidth_lower_um"].values[0]

            # grid_size_top_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_upper_um"].values[0]
            # grid_size_top_t1 = df_sample.loc[df_sample["leaf stage"] == "J0", "cellwidth_upper_um"].values[0]
            # grid_size_bot_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_lower_um"].values[0]
            # grid_size_bot_t1 = df_sample.loc[df_sample["leaf stage"] == "J0", "cellwidth_lower_um"].values[0]

            phi_top = grid_size_top_t1 / grid_size_top_t0
            phi_bot = grid_size_bot_t1 / grid_size_bot_t0

            j0_series = pd.DataFrame({"sample": [sample], "leaf stage": "J0", "ratio_top": [phi_top], "ratio_bot": [phi_bot]})
        if "J1" in df_sample["leaf stage"].values:
             # gridsize refers to the variable epsilon above
            grid_size_top_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellnumb_upper"].values[0] * df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_upper_um"].values[0]
            grid_size_top_t1 = df_sample.loc[df_sample["leaf stage"] == "J1", "cellnumb_upper"].values[0] * df_sample.loc[df_sample["leaf stage"] == "J1", "cellwidth_upper_um"].values[0]
            grid_size_bot_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellnumb_lower"].values[0] * df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_lower_um"].values[0]
            grid_size_bot_t1 = df_sample.loc[df_sample["leaf stage"] == "J1", "cellnumb_lower"].values[0] * df_sample.loc[df_sample["leaf stage"] == "J1", "cellwidth_lower_um"].values[0]

            # grid_size_top_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_upper_um"].values[0]
            # grid_size_top_t1 = df_sample.loc[df_sample["leaf stage"] == "J1", "cellwidth_upper_um"].values[0]
            # grid_size_bot_t0 = df_sample.loc[df_sample["leaf stage"] == "A", "cellwidth_lower_um"].values[0]
            # grid_size_bot_t1 = df_sample.loc[df_sample["leaf stage"] == "J1", "cellwidth_lower_um"].values[0]

            phi_top = grid_size_top_t1 / grid_size_top_t0
            phi_bot = grid_size_bot_t1 / grid_size_bot_t0

            j1_series = pd.DataFrame({"sample": [sample], "leaf stage": "J1", "ratio_top": [phi_top], "ratio_bot": [phi_bot]})

        return pd.concat([x for x in [j0_series, j1_series] if x is not None])


df_mismatch_data = df_mismatch_data.groupby("sample").apply(_apply_calculate_mismatch_ratios).reset_index(drop=True)
df_mismatch_data

C:\Users\nicks\AppData\Local\Temp\ipykernel_21696\3653896839.py:42: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df_mismatch_data = df_mismatch_data.groupby("sample").apply(_apply_calculate_mismatch_ratios).reset_index(drop=True)


sample leaf stage  ratio_top  ratio_bot
0    C01         J0   0.667032   1.108861
1    C02         J0   0.626911   0.684462
2    C02         J1   0.414668   0.704043
3    C03         J0   0.385149   0.785947
4    C03         J1   0.412980   0.815735
5    C04         J0   0.450032   1.018871
6    C04         J1   1.004013   2.515442

### 10. Incorperating the cell dimensions data into the first dataset

In [219]:
df_radius_from_measured_mismatch = df.copy()
df_radius_from_measured_mismatch = df_radius_from_measured_mismatch.merge(df_mismatch_data, on=["sample", "leaf stage"])

Finally we need to convert the ratios into layer mismatche. I propose the following relation
$$\frac{1}{1+hc_{R}B} \overset{!}{=} \phi \iff B = \frac{\frac{1}{\phi} - 1}{c_Rh}$$
where $B$ refers to $B_{\text{top}}$, $B_{\text{bot}}$ and $\phi$ refers to $\phi_{\text{top}}$, $\phi_{\text{bot}}$ and $c_R$ is a translation parameter we introduce to translate the microsopic structure ratio to an actual layer mismatch. Here we choose $c_R = 12.5$ for all samples below.

This translation parameter can be understood as follows: If in our leafs the cells grow, or new cells form this likely would produce a layer mismatch but the relationship between the number of cells spawning and the layer mismatch is likely not the identity (cells spawning is not the same as the distance between atoms), but this relationship is likely strongly positively correlated (more cells in one layer implies more mismatch). $c_R$ determines the slope of this relationship.

Alternatively to the above we could also use something like
$$\frac{1}{1+hB} \overset{!}{=} c_R \phi$$

For reference, *Bernd Schmidt* in his paper used the relation
    $$\frac{1+hB_{\text{top}}}{1+hB_{\text{bot}}} = \frac{\alpha_\text{bot}}{\alpha_\text{top}}, \text{i.e.} \qquad \frac{1}{1+hB_{\text{bot}}} = \alpha_\text{bot}, \text{ and } \frac{1}{1+hB_{\text{top}}} = \alpha_\text{top}$$

In [220]:
c_R = 11.3

df_radius_from_measured_mismatch["B_bot"] = (1 / df_radius_from_measured_mismatch["ratio_bot"] - 1) / df_radius_from_measured_mismatch["plate thickness h_mm"] / c_R
df_radius_from_measured_mismatch["B_top"] = (1 / df_radius_from_measured_mismatch["ratio_top"] - 1) / df_radius_from_measured_mismatch["plate thickness h_mm"] / c_R

df_radius_from_measured_mismatch.drop(["ratio_top", "ratio_bot"], axis=1, inplace=True)
df_radius_from_measured_mismatch

leaf stage sample                                   full sample name   a_mm  \
0         J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227   
1         J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.948   
2         J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  0.896   
3         J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  1.387   
4         J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590   
5         J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156   
6         J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242   

    s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0  3.538  1.888708      0.529462                  0.057000   
1  3.553  2.138532      0.467611                  0.060667   
2  2.754  1.506108      0.663963                  0.051367   
3  4.405  2.442241      0.409460                  0.050667   
4  2.394  1.509245      0.662583                  0.054667   
5  4.176  2.463702      0.405893                  0.063333   
6  3.804  2.077362      0.481380                  0.054000   

   thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                  0.108667          0.155936            0.165667   
1                  0.143433          0.202760            0.204100   
2                  0.171333          0.269346            0.222700   
3                  0.131333          0.221612            0.182000   
4                  0.106333          0.160455            0.161000   
5                  0.142667          0.192557            0.206000   
6                  0.149333          0.234426            0.203333   

   leaf_width_mm  plate thickness h_mm      B_bot      B_top  
0         60.448              0.002741  -3.170038  16.118452  
1         73.546              0.002775  14.700783  18.977688  
2         41.638              0.005348   4.506302  26.413905  
3         46.339              0.003928  -0.417324  27.535401  
4         53.834              0.002991  12.438892  41.768991  
5         43.277              0.004760   4.199562  26.426289  
6         64.786              0.003139 -16.987098  -0.112692

#### 11. Calculating the Optimal Radius from the Measured Mismatch Data

Finally we use the `tubes.py` module to calculate the optimal radius. I tried out different values for the Young's Modulus and Poisson's Ratio.

In [221]:
def _calculate_radius_with_measured_layer_mismatch(df_row):
    output = optimal_radius(tau=df_row["middle surface τ"],
                            h=df_row["plate thickness h_mm"],
                            total_thickness=df_row["total thickness_mm"],
                            btop=df_row["B_top"],
                            bbot=df_row["B_bot"],
                            youngs_modulus=[YOUNGS_MODULUS * 1.09, YOUNGS_MODULUS],
                            poisson_ratio=[0.4, 0.4],
                            measured_radius=df_row["r_mm"])

    return pd.Series({"calculated_radius": output["calculated_radius"], "relative_deviation": str(round(100 * output["relative_deviation"], 2)) + "%"})


df_radius_from_measured_mismatch_merged = df_radius_from_measured_mismatch.join(df_radius_from_measured_mismatch.apply(_calculate_radius_with_measured_layer_mismatch, axis=1).reset_index(drop=True))
df_radius_from_measured_mismatch_merged

leaf stage sample                                   full sample name   a_mm  \
0         J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227   
1         J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.948   
2         J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  0.896   
3         J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  1.387   
4         J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590   
5         J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156   
6         J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242   

    s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0  3.538  1.888708      0.529462                  0.057000   
1  3.553  2.138532      0.467611                  0.060667   
2  2.754  1.506108      0.663963                  0.051367   
3  4.405  2.442241      0.409460                  0.050667   
4  2.394  1.509245      0.662583                  0.054667   
5  4.176  2.463702      0.405893                  0.063333   
6  3.804  2.077362      0.481380                  0.054000   

   thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                  0.108667          0.155936            0.165667   
1                  0.143433          0.202760            0.204100   
2                  0.171333          0.269346            0.222700   
3                  0.131333          0.221612            0.182000   
4                  0.106333          0.160455            0.161000   
5                  0.142667          0.192557            0.206000   
6                  0.149333          0.234426            0.203333   

   leaf_width_mm  plate thickness h_mm      B_bot      B_top  \
0         60.448              0.002741  -3.170038  16.118452   
1         73.546              0.002775  14.700783  18.977688   
2         41.638              0.005348   4.506302  26.413905   
3         46.339              0.003928  -0.417324  27.535401   
4         53.834              0.002991  12.438892  41.768991   
5         43.277              0.004760   4.199562  26.426289   
6         64.786              0.003139 -16.987098  -0.112692   

   calculated_radius relative_deviation  
0           1.702647             -9.85%  
1          14.856881            594.72%  
2           1.365786             -9.32%  
3           1.028172             -57.9%  
4           1.058960            -29.84%  
5           1.159000            -52.96%  
6           2.253485              8.48%

### 12. Exporting the Results and Calculating Statistical Parameters

In [222]:
df_radius_from_measured_mismatch_merged.to_excel("results_28_11_2024_nick.xlsx", index=False)
print("Radius Statistics")
radius_stats = pd.Series({
    "correlation": df_radius_from_measured_mismatch_merged["r_mm"].corr(df_radius_from_measured_mismatch_merged["calculated_radius"]),
    "variance of differences": (df_radius_from_measured_mismatch_merged["r_mm"] - df_radius_from_measured_mismatch_merged["calculated_radius"]).var(),
    "variance of measured": df_radius_from_measured_mismatch_merged["r_mm"].var(),
    "mean of differences": (df_radius_from_measured_mismatch_merged["r_mm"] - df_radius_from_measured_mismatch_merged["calculated_radius"]).mean()
})

radius_stats

Radius Statistics


correlation                 0.145558
variance of differences    25.521011
variance of measured        0.155472
mean of differences        -1.342719
dtype: float64

### 13. Discussion of the Results and Notes

The results show that the order of magnitude of the radius is captured correctly. However this is mainly due to the "translation parameter" $c_R$ that scales the atomistic layer mismatch $B$ compared to the ratio $\phi$. This parameter is chosen such that the correct order of magnitude is captured. The model is not able to capture fine nuances between samples (for the given data). For example in the table above we see a larger radius for probe 3 copared to probe 2, while the model predicts the opposite, i.e the correlation between measured radii and calculated radii is not too high. Strictly speaking the correlation is not the correct statistic here since the samples are not "equally distributed" but gives a general indication of the goodness of fit.

The results are unfortunately not as good as anticipated after the first results. However since we assumed the material to be isotropic and also the strain between layers cannot be measured accurately, anything better than the above would be too good to be true.

What stands out is probe 1 where the relative deviation is alot higher that that of the rest. This is because if you look at B_bot and B_top they have the same sign and they are pretty close to another i.e not alot of layer mismatch i.e. large radius.

There were several different types of measurements made, for example "counted on whole leaf or 1 leaf half?" or the different time stages J0 / J1 compared to A. In the calculation above I treated all datapoints as equal and used them as is. However it is unclear whether different types of datapoints are comparable.


# __12.01.2025__ 

After the discussing the results, we decided to no longer regard the cell size data of the same C0X samples at different time stages but instead use the average at each time sample i.e instead of using the measured sizes for C01 at time stages A and J0, we instead use the averaged sizes of all probes at time stage A and J0 respecitely. The following table contains these averaged values.

The goal of this section is to perform the exact same calculation as above and replacing the grid ratio's with the averaged grid ratio's:

In [223]:
df_cell_data_averages = pd.read_excel("pilea data 4.0/Spod_statistics.xlsx")
df_cell_data_averages

leaf.stage  mean_cellwidth_upper  median_cellwidth_upper  \
0        J-1                 10.37                    9.98   
1         J0                 21.45                   22.15   
2         J1                 27.86                   26.11   
3          A                 36.76                   37.41   

   mean_cellwidth_lower  median_cellwidth_lower  mean_cellthickn_upper  \
0                  9.35                    8.80                  15.63   
1                 24.80                   25.56                  22.87   
2                 30.68                   27.73                  29.63   
3                 31.14                   30.51                  33.69   

   median_cellthickn_upper  mean_cellthickn_lower  median_cellthickn_lower  \
0                    15.51                  13.33                    24.88   
1                    23.22                  20.80                    13.37   
2                    26.77                  22.46                    21.89   
3                    32.94                  25.32                    21.12   

   mean_cellnumb_upper  median_cellnumb_upper  mean_cellnumb_lower  \
0               1472.4                   1603               1905.4   
1               2391.8                   2354               2768.4   
2               2222.0                   2076               2688.0   
3               2478.2                   2209               2389.8   

   median_cellnumb_lower  
0                   1962  
1                   2850  
2                   2626  
3                   2160

#### 14. Calculate averaged Ratios

We again calculate the so-called "grid-size" completely analog to above, just instead with the averaged values:

In [224]:
# gridsize refers to the variable epsilon above
grid_size_top_t0 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "A", "mean_cellnumb_upper"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "A", "mean_cellwidth_upper"].values[0]
grid_size_top_t1 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J0", "mean_cellnumb_upper"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J0", "mean_cellwidth_upper"].values[0]
grid_size_top_t2 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J1", "mean_cellnumb_upper"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J1", "mean_cellwidth_upper"].values[0]

grid_size_bot_t0 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "A", "mean_cellnumb_lower"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "A", "mean_cellwidth_lower"].values[0]
grid_size_bot_t1 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J0", "mean_cellnumb_lower"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J0", "mean_cellwidth_lower"].values[0]
grid_size_bot_t2 = df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J1", "mean_cellnumb_lower"].values[0] * df_cell_data_averages.loc[df_cell_data_averages["leaf.stage"] == "J1", "mean_cellwidth_lower"].values[0]


phi_top_J0A = grid_size_top_t1 / grid_size_top_t0
phi_bot_J0A = grid_size_bot_t1 / grid_size_bot_t0

phi_top_J1A = grid_size_top_t2 / grid_size_top_t0
phi_bot_J1A = grid_size_bot_t2 / grid_size_bot_t0

Now we again use the table containing the radius measurements and extend it with the averaged ratios calcuklated above:

In [225]:
df_radius_from_measured_mismatch_averaged = df.copy()
#df_radius_from_measured_mismatch = df_radius_from_measured_mismatch.merge(df_mismatch_data, on=["sample", "leaf stage"])

df_radius_from_measured_mismatch_averaged.loc[df_radius_from_measured_mismatch_averaged["leaf stage"] == "J0", "ratio_bot"] = phi_bot_J0A
df_radius_from_measured_mismatch_averaged.loc[df_radius_from_measured_mismatch_averaged["leaf stage"] == "J0", "ratio_top"] = phi_top_J0A
df_radius_from_measured_mismatch_averaged.loc[df_radius_from_measured_mismatch_averaged["leaf stage"] == "J1", "ratio_bot"] = phi_bot_J1A
df_radius_from_measured_mismatch_averaged.loc[df_radius_from_measured_mismatch_averaged["leaf stage"] == "J1", "ratio_top"] = phi_top_J1A

df_radius_from_measured_mismatch_averaged

leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.155936            0.165667   
1                   0.143433          0.202760            0.204100   
2                   0.171333          0.269346            0.222700   
3                   0.131333          0.221612            0.182000   
4                   0.144667          0.203404            0.205667   
5                   0.106000          0.101134            0.176333   
6                   0.106333          0.160455            0.161000   
7                   0.142667          0.192557            0.206000   
8                   0.149333          0.234426            0.203333   
9                   0.136333          0.230357            0.186667   
10                  0.131333          0.208633            0.185333   

    leaf_width_mm  plate thickness h_mm  ratio_bot  ratio_top  
0          60.448              0.002741   0.922572   0.563171  
1          73.546              0.002775   0.922572   0.563171  
2          41.638              0.005348   0.922572   0.563171  
3          46.339              0.003928   0.922572   0.563171  
4          62.311              0.003301   0.922572   0.563171  
5          60.348              0.002922   1.108165   0.679537  
6          53.834              0.002991   1.108165   0.679537  
7          43.277              0.004760   1.108165   0.679537  
8          64.786              0.003139   1.108165   0.679537  
9          81.568              0.002288   1.108165   0.679537  
10         45.794              0.004047   1.108165   0.679537

#### 15. Deriving averaged Layer Mismatches

Using the radios we can now derive a layer mismatch same as above, just instead with the averaged ratios. Here we again use the same parameter $c_R$:

In [226]:
c_R = 11.3

df_radius_from_measured_mismatch_averaged["B_bot"] = (1 / df_radius_from_measured_mismatch_averaged["ratio_bot"] - 1) / df_radius_from_measured_mismatch_averaged["plate thickness h_mm"] / c_R
df_radius_from_measured_mismatch_averaged["B_top"] = (1 / df_radius_from_measured_mismatch_averaged["ratio_top"] - 1) / df_radius_from_measured_mismatch_averaged["plate thickness h_mm"] / c_R

df_radius_from_measured_mismatch_averaged.drop(["ratio_top", "ratio_bot"], axis=1, inplace=True)
df_radius_from_measured_mismatch_averaged

leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.155936            0.165667   
1                   0.143433          0.202760            0.204100   
2                   0.171333          0.269346            0.222700   
3                   0.131333          0.221612            0.182000   
4                   0.144667          0.203404            0.205667   
5                   0.106000          0.101134            0.176333   
6                   0.106333          0.160455            0.161000   
7                   0.142667          0.192557            0.206000   
8                   0.149333          0.234426            0.203333   
9                   0.136333          0.230357            0.186667   
10                  0.131333          0.208633            0.185333   

    leaf_width_mm  plate thickness h_mm     B_bot      B_top  
0          60.448              0.002741  2.709973  25.046064  
1          73.546              0.002775  2.676297  24.734819  
2          41.638              0.005348  1.388634  12.834009  
3          46.339              0.003928  1.891008  17.477043  
4          62.311              0.003301  2.250189  20.796656  
5          60.348              0.002922 -2.956192  14.282827  
6          53.834              0.002991 -2.888251  13.954571  
7          43.277              0.004760 -1.814655   8.767497  
8          64.786              0.003139 -2.752179  13.297142  
9          81.568              0.002288 -3.774480  18.236385  
10         45.794              0.004047 -2.134318  10.311949

#### 16. Calculating radii from layer mismatches derived by averaged values

Finally we again calculate the anticipated radius using `tubes.py`. I removed the factor $1.09$ for the *Young's Modulus* of the bottom layer as there was not a real foundation putting it there in the first place.

In [227]:
def _calculate_radius_with_measured_layer_mismatch(df_row):
    output = optimal_radius(tau=df_row["middle surface τ"],
                            h=df_row["plate thickness h_mm"],
                            total_thickness=df_row["total thickness_mm"],
                            btop=df_row["B_top"],
                            bbot=df_row["B_bot"],
                            youngs_modulus=[YOUNGS_MODULUS, YOUNGS_MODULUS],
                            poisson_ratio=[0.4, 0.4],
                            measured_radius=df_row["r_mm"])

    return pd.Series({"calculated_radius": output["calculated_radius"], "relative_deviation": str(round(100 * output["relative_deviation"], 2)) + "%"})


df_radius_from_measured_mismatch_averaged_merged = df_radius_from_measured_mismatch_averaged.join(df_radius_from_measured_mismatch_averaged.apply(_calculate_radius_with_measured_layer_mismatch, axis=1).reset_index(drop=True))
df_radius_from_measured_mismatch_averaged_merged

leaf stage sample                                   full sample name  \
0          J0    C01  2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...   
1          J0    C02  2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...   
2          J0    C03          SPJO CO3 PL1 Schnitt2 2x zus_freigestellt   
3          J0    C04   SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt   
4          J0    C06          SPJO CO6 PL1 Schnitt3 2x zus_freigestellt   
5          J1    A01  2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...   
6          J1    C02  2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...   
7          J1    C03            SP J1 CO3 PL5 Schnitt 2 2x_freigestellt   
8          J1    C04  SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...   
9          J1    C05           SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt   
10         J1    C07         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt   

     a_mm   s_mm      r_mm  curvature_mm  thickness upper layer_mm  \
0   1.227  3.538  1.888708      0.529462                  0.057000   
1   0.948  3.553  2.138532      0.467611                  0.060667   
2   0.896  2.754  1.506108      0.663963                  0.051367   
3   1.387  4.405  2.442241      0.409460                  0.050667   
4   1.204  3.987  2.252350      0.443981                  0.061000   
5   1.252  3.526  1.867282      0.535538                  0.070333   
6   0.590  2.394  1.509245      0.662583                  0.054667   
7   1.156  4.176  2.463702      0.405893                  0.063333   
8   1.242  3.804  2.077362      0.481380                  0.054000   
9   0.704  3.790  2.902444      0.344537                  0.050333   
10  1.246  4.742  2.878875      0.347358                  0.054000   

    thickness lower layer_mm  middle surface τ  total thickness_mm  \
0                   0.108667          0.155936            0.165667   
1                   0.143433          0.202760            0.204100   
2                   0.171333          0.269346            0.222700   
3                   0.131333          0.221612            0.182000   
4                   0.144667          0.203404            0.205667   
5                   0.106000          0.101134            0.176333   
6                   0.106333          0.160455            0.161000   
7                   0.142667          0.192557            0.206000   
8                   0.149333          0.234426            0.203333   
9                   0.136333          0.230357            0.186667   
10                  0.131333          0.208633            0.185333   

    leaf_width_mm  plate thickness h_mm     B_bot      B_top  \
0          60.448              0.002741  2.709973  25.046064   
1          73.546              0.002775  2.676297  24.734819   
2          41.638              0.005348  1.388634  12.834009   
3          46.339              0.003928  1.891008  17.477043   
4          62.311              0.003301  2.250189  20.796656   
5          60.348              0.002922 -2.956192  14.282827   
6          53.834              0.002991 -2.888251  13.954571   
7          43.277              0.004760 -1.814655   8.767497   
8          64.786              0.003139 -2.752179  13.297142   
9          81.568              0.002288 -3.774480  18.236385   
10         45.794              0.004047 -2.134318  10.311949   

    calculated_radius relative_deviation  
0            1.427561            -24.42%  
1            1.900155            -11.15%  
2            2.440606             62.05%  
3            1.761884            -27.86%  
4            1.917140            -14.88%  
5            1.738092             -6.92%  
6            1.696767             12.42%  
7            2.286564             -7.19%  
8            2.463836              18.6%  
9            2.240161            -22.82%  
10           2.121429            -26.31%

#### 17. Statisticts for averaged calculation

In [228]:
df_radius_from_measured_mismatch_averaged_merged.to_excel("results_12_01_2025_nick.xlsx", index=False)
print("Radius Statistics")
radius_stats_averaged = pd.Series({
    "correlation": df_radius_from_measured_mismatch_averaged_merged["r_mm"].corr(df_radius_from_measured_mismatch_averaged_merged["calculated_radius"]),
    "variance of differences": (df_radius_from_measured_mismatch_averaged_merged["r_mm"] - df_radius_from_measured_mismatch_averaged_merged["calculated_radius"]).var(),
    "variance of measured": df_radius_from_measured_mismatch_averaged_merged["r_mm"].var(),
    "mean of differences": (df_radius_from_measured_mismatch_averaged_merged["r_mm"] - df_radius_from_measured_mismatch_averaged_merged["calculated_radius"]).mean()
})

radius_stats_averaged

Radius Statistics


correlation                0.239858
variance of differences    0.262340
variance of measured       0.226353
mean of differences        0.175696
dtype: float64

#### 18. Discussion of the averaged results

The results are actually interesting! We used the same parameter $c_R$ and the deviations are really small! Of course $c_R$ was determined using these samples, but this may indicate that this constant $c_R$ is of a more general relevance for this setting. Also the correlation is pretty high! It would be interesting to perform the same calculation above with a different set of samples but with the same $c_R$.

I will take a closer look at whether the correlation is the correct statistic here. As discussed I am sceptical but I haven't formulated that in a rigorous way. This will follow.

# **22.01.2025 (corrected 03.02.2025)**


In [229]:
df_radius_from_measured_mismatch_averaged_final = pd.read_excel("pilea data 5.0/Spod_leaf-curvature_final-measurements_headers-deleted.xlsx", sheet_name="Data")

# We again translate the middle surface as explained above
df_radius_from_measured_mismatch_averaged_final["middle surface τ"] = 0.5 * (1 - df_radius_from_measured_mismatch_averaged_final["middle surface τ"]) / (df_radius_from_measured_mismatch_averaged_final["middle surface τ"] + 1)

df_radius_from_measured_mismatch_averaged_final

leaf stage sample leaf half  \
0          J0    C01     inner   
1          J0    C02     inner   
2          J0    C03     inner   
3          J0    C04     inner   
4          J0    C06     inner   
5          J1    A01     inner   
6          J1    C02     inner   
7          J1    C03     inner   
8          J1    C04     inner   
9          J1    C05     inner   
10         J1    C07     inner   

                                     full sample name   a_mm   s_mm      r_mm  \
0   2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227  3.538  1.888708   
1   2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.860  2.563  1.317571   
2           SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  1.191  2.380  1.264500   
3    SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  0.656  2.481  1.262182   
4           SPJO CO6 PL1 Schnitt3 2x zus_freigestellt  1.415  3.109  1.635001   
5   2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...  1.252  3.526  1.867282   
6   2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590  2.394  1.509245   
7             SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156  4.176  2.463702   
8   SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242  3.804  2.077362   
9            SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt  0.704  3.790  2.902444   
10         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt  1.246  4.742  2.878875   

    curvature_mm  thickness upper layer_mm  thickness lower layer_mm  \
0       0.529462                  0.057000                  0.108667   
1       0.758972                  0.060667                  0.143433   
2       0.790826                  0.051367                  0.171333   
3       0.792279                  0.050667                  0.131333   
4       0.611620                  0.061000                  0.144667   
5       0.535538                  0.070333                  0.106000   
6       0.662583                  0.054667                  0.106333   
7       0.405893                  0.063333                  0.142667   
8       0.481380                  0.054000                  0.149333   
9       0.344537                  0.050333                  0.136333   
10      0.347358                  0.054000                  0.131333   

    middle surface τ  total thickness_mm  leaf_width_mm  plate thickness h_mm  
0           0.155936            0.165667         60.448              0.002741  
1           0.202760            0.204100         73.546              0.002775  
2           0.269346            0.222700         41.638              0.005348  
3           0.221612            0.182000         46.339              0.003928  
4           0.203404            0.205667         62.311              0.003301  
5           0.101134            0.176333         60.348              0.002922  
6           0.160455            0.161000         53.834              0.002991  
7           0.192557            0.206000         43.277              0.004760  
8           0.234426            0.203333         64.786              0.003139  
9           0.230357            0.186667         81.568              0.002288  
10          0.208633            0.185333         45.794              0.004047

In [230]:
df_radius_from_measured_mismatch_averaged_final.loc[df_radius_from_measured_mismatch_averaged_final["leaf stage"] == "J0", "ratio_bot"] = phi_bot_J0A
df_radius_from_measured_mismatch_averaged_final.loc[df_radius_from_measured_mismatch_averaged_final["leaf stage"] == "J0", "ratio_top"] = phi_top_J0A
df_radius_from_measured_mismatch_averaged_final.loc[df_radius_from_measured_mismatch_averaged_final["leaf stage"] == "J1", "ratio_bot"] = phi_bot_J1A
df_radius_from_measured_mismatch_averaged_final.loc[df_radius_from_measured_mismatch_averaged_final["leaf stage"] == "J1", "ratio_top"] = phi_top_J1A

df_radius_from_measured_mismatch_averaged_final

leaf stage sample leaf half  \
0          J0    C01     inner   
1          J0    C02     inner   
2          J0    C03     inner   
3          J0    C04     inner   
4          J0    C06     inner   
5          J1    A01     inner   
6          J1    C02     inner   
7          J1    C03     inner   
8          J1    C04     inner   
9          J1    C05     inner   
10         J1    C07     inner   

                                     full sample name   a_mm   s_mm      r_mm  \
0   2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227  3.538  1.888708   
1   2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.860  2.563  1.317571   
2           SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  1.191  2.380  1.264500   
3    SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  0.656  2.481  1.262182   
4           SPJO CO6 PL1 Schnitt3 2x zus_freigestellt  1.415  3.109  1.635001   
5   2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...  1.252  3.526  1.867282   
6   2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590  2.394  1.509245   
7             SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156  4.176  2.463702   
8   SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242  3.804  2.077362   
9            SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt  0.704  3.790  2.902444   
10         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt  1.246  4.742  2.878875   

    curvature_mm  thickness upper layer_mm  thickness lower layer_mm  \
0       0.529462                  0.057000                  0.108667   
1       0.758972                  0.060667                  0.143433   
2       0.790826                  0.051367                  0.171333   
3       0.792279                  0.050667                  0.131333   
4       0.611620                  0.061000                  0.144667   
5       0.535538                  0.070333                  0.106000   
6       0.662583                  0.054667                  0.106333   
7       0.405893                  0.063333                  0.142667   
8       0.481380                  0.054000                  0.149333   
9       0.344537                  0.050333                  0.136333   
10      0.347358                  0.054000                  0.131333   

    middle surface τ  total thickness_mm  leaf_width_mm  plate thickness h_mm  \
0           0.155936            0.165667         60.448              0.002741   
1           0.202760            0.204100         73.546              0.002775   
2           0.269346            0.222700         41.638              0.005348   
3           0.221612            0.182000         46.339              0.003928   
4           0.203404            0.205667         62.311              0.003301   
5           0.101134            0.176333         60.348              0.002922   
6           0.160455            0.161000         53.834              0.002991   
7           0.192557            0.206000         43.277              0.004760   
8           0.234426            0.203333         64.786              0.003139   
9           0.230357            0.186667         81.568              0.002288   
10          0.208633            0.185333         45.794              0.004047   

    ratio_bot  ratio_top  
0    0.922572   0.563171  
1    0.922572   0.563171  
2    0.922572   0.563171  
3    0.922572   0.563171  
4    0.922572   0.563171  
5    1.108165   0.679537  
6    1.108165   0.679537  
7    1.108165   0.679537  
8    1.108165   0.679537  
9    1.108165   0.679537  
10   1.108165   0.679537

In [231]:
c_R = 10.8

df_radius_from_measured_mismatch_averaged_final["B_bot"] = (1 / df_radius_from_measured_mismatch_averaged_final["ratio_bot"] - 1) / df_radius_from_measured_mismatch_averaged_final["plate thickness h_mm"] / c_R
df_radius_from_measured_mismatch_averaged_final["B_top"] = (1 / df_radius_from_measured_mismatch_averaged_final["ratio_top"] - 1) / df_radius_from_measured_mismatch_averaged_final["plate thickness h_mm"] / c_R

df_radius_from_measured_mismatch_averaged_final.drop(["ratio_top", "ratio_bot"], axis=1, inplace=True)
df_radius_from_measured_mismatch_averaged_final

leaf stage sample leaf half  \
0          J0    C01     inner   
1          J0    C02     inner   
2          J0    C03     inner   
3          J0    C04     inner   
4          J0    C06     inner   
5          J1    A01     inner   
6          J1    C02     inner   
7          J1    C03     inner   
8          J1    C04     inner   
9          J1    C05     inner   
10         J1    C07     inner   

                                     full sample name   a_mm   s_mm      r_mm  \
0   2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227  3.538  1.888708   
1   2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.860  2.563  1.317571   
2           SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  1.191  2.380  1.264500   
3    SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  0.656  2.481  1.262182   
4           SPJO CO6 PL1 Schnitt3 2x zus_freigestellt  1.415  3.109  1.635001   
5   2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...  1.252  3.526  1.867282   
6   2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590  2.394  1.509245   
7             SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156  4.176  2.463702   
8   SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242  3.804  2.077362   
9            SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt  0.704  3.790  2.902444   
10         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt  1.246  4.742  2.878875   

    curvature_mm  thickness upper layer_mm  thickness lower layer_mm  \
0       0.529462                  0.057000                  0.108667   
1       0.758972                  0.060667                  0.143433   
2       0.790826                  0.051367                  0.171333   
3       0.792279                  0.050667                  0.131333   
4       0.611620                  0.061000                  0.144667   
5       0.535538                  0.070333                  0.106000   
6       0.662583                  0.054667                  0.106333   
7       0.405893                  0.063333                  0.142667   
8       0.481380                  0.054000                  0.149333   
9       0.344537                  0.050333                  0.136333   
10      0.347358                  0.054000                  0.131333   

    middle surface τ  total thickness_mm  leaf_width_mm  plate thickness h_mm  \
0           0.155936            0.165667         60.448              0.002741   
1           0.202760            0.204100         73.546              0.002775   
2           0.269346            0.222700         41.638              0.005348   
3           0.221612            0.182000         46.339              0.003928   
4           0.203404            0.205667         62.311              0.003301   
5           0.101134            0.176333         60.348              0.002922   
6           0.160455            0.161000         53.834              0.002991   
7           0.192557            0.206000         43.277              0.004760   
8           0.234426            0.203333         64.786              0.003139   
9           0.230357            0.186667         81.568              0.002288   
10          0.208633            0.185333         45.794              0.004047   

       B_bot      B_top  
0   2.835435  26.205604  
1   2.800199  25.879950  
2   1.452923  13.428176  
3   1.978555  18.286166  
4   2.354365  21.759464  
5  -3.093052  14.944069  
6  -3.021966  14.600616  
7  -1.898667   9.173400  
8  -2.879595  13.912750  
9  -3.949225  19.080662  
10 -2.233129  10.789354

In [232]:
def _calculate_radius_with_measured_layer_mismatch(df_row):
    output = optimal_radius(tau=df_row["middle surface τ"],
                            h=df_row["plate thickness h_mm"],
                            total_thickness=df_row["total thickness_mm"],
                            btop=df_row["B_top"],
                            bbot=df_row["B_bot"],
                            youngs_modulus=[YOUNGS_MODULUS, YOUNGS_MODULUS],
                            poisson_ratio=[0.4, 0.4],
                            measured_radius=df_row["r_mm"])

    return pd.Series({"calculated_radius": output["calculated_radius"], "relative_deviation": str(round(100 * output["relative_deviation"], 2)) + "%"})


df_radius_from_measured_mismatch_averaged_final_merged = df_radius_from_measured_mismatch_averaged_final.join(df_radius_from_measured_mismatch_averaged_final.apply(_calculate_radius_with_measured_layer_mismatch, axis=1).reset_index(drop=True))
df_radius_from_measured_mismatch_averaged_final_merged

leaf stage sample leaf half  \
0          J0    C01     inner   
1          J0    C02     inner   
2          J0    C03     inner   
3          J0    C04     inner   
4          J0    C06     inner   
5          J1    A01     inner   
6          J1    C02     inner   
7          J1    C03     inner   
8          J1    C04     inner   
9          J1    C05     inner   
10         J1    C07     inner   

                                     full sample name   a_mm   s_mm      r_mm  \
0   2024-02-13_SP_JO_CO1_PL13_Schnitt2_5µ_4x-zus__...  1.227  3.538  1.888708   
1   2024-02-02_Spod_JO_CO2_PL14_Schnitt1_Bild1_5µ_...  0.860  2.563  1.317571   
2           SPJO CO3 PL1 Schnitt2 2x zus_freigestellt  1.191  2.380  1.264500   
3    SP JO CO4 PL2 Schnitt2 Bild1 2x zus_freigestellt  0.656  2.481  1.262182   
4           SPJO CO6 PL1 Schnitt3 2x zus_freigestellt  1.415  3.109  1.635001   
5   2024-02-15_SP_J1_AO1_PL_11_Schnitt2_2x zus_fre...  1.252  3.526  1.867282   
6   2024-01-29_Spod_J1_C02_PL7_Schnitt1_4x zus_fre...  0.590  2.394  1.509245   
7             SP J1 CO3 PL5 Schnitt 2 2x_freigestellt  1.156  4.176  2.463702   
8   SP J1 CO4 PL1 Schnitt2 ganzes Bild 4x zus_frei...  1.242  3.804  2.077362   
9            SPJ1CO5 PL2 Schnitt2 2x zus_freigestellt  0.704  3.790  2.902444   
10         SP J1 CO7 PL2 Schnitt2 4x zus_freigestellt  1.246  4.742  2.878875   

    curvature_mm  thickness upper layer_mm  thickness lower layer_mm  \
0       0.529462                  0.057000                  0.108667   
1       0.758972                  0.060667                  0.143433   
2       0.790826                  0.051367                  0.171333   
3       0.792279                  0.050667                  0.131333   
4       0.611620                  0.061000                  0.144667   
5       0.535538                  0.070333                  0.106000   
6       0.662583                  0.054667                  0.106333   
7       0.405893                  0.063333                  0.142667   
8       0.481380                  0.054000                  0.149333   
9       0.344537                  0.050333                  0.136333   
10      0.347358                  0.054000                  0.131333   

    middle surface τ  total thickness_mm  leaf_width_mm  plate thickness h_mm  \
0           0.155936            0.165667         60.448              0.002741   
1           0.202760            0.204100         73.546              0.002775   
2           0.269346            0.222700         41.638              0.005348   
3           0.221612            0.182000         46.339              0.003928   
4           0.203404            0.205667         62.311              0.003301   
5           0.101134            0.176333         60.348              0.002922   
6           0.160455            0.161000         53.834              0.002991   
7           0.192557            0.206000         43.277              0.004760   
8           0.234426            0.203333         64.786              0.003139   
9           0.230357            0.186667         81.568              0.002288   
10          0.208633            0.185333         45.794              0.004047   

       B_bot      B_top  calculated_radius relative_deviation  
0   2.835435  26.205604           1.364394            -27.76%  
1   2.800199  25.879950           1.816078             37.84%  
2   1.452923  13.428176           2.332615             84.47%  
3   1.978555  18.286166           1.683924             33.41%  
4   2.354365  21.759464           1.832311             12.07%  
5  -3.093052  14.944069           1.661186            -11.04%  
6  -3.021966  14.600616           1.621689              7.45%  
7  -1.898667   9.173400           2.185389             -11.3%  
8  -2.879595  13.912750           2.354817             13.36%  
9  -3.949225  19.080662           2.141039            -26.23%  
10 -2.233129  10.789354           2.027561            -29.57%

In [233]:
df_radius_from_measured_mismatch_averaged_final_merged.to_excel("results_22_01_2025_nick.xlsx", index=False)
print("Radius Statistics")
radius_stats_averaged = pd.Series({
    "correlation": df_radius_from_measured_mismatch_averaged_final_merged["r_mm"].corr(df_radius_from_measured_mismatch_averaged_final_merged["calculated_radius"]),
    "variance of differences": (df_radius_from_measured_mismatch_averaged_final_merged["r_mm"] - df_radius_from_measured_mismatch_averaged_final_merged["calculated_radius"]).var(),
    "variance of measured": df_radius_from_measured_mismatch_averaged_final_merged["r_mm"].var(),
    "mean of differences": (df_radius_from_measured_mismatch_averaged_final_merged["r_mm"] - df_radius_from_measured_mismatch_averaged_final_merged["calculated_radius"]).mean()
})

radius_stats_averaged

Radius Statistics


correlation                0.338367
variance of differences    0.339492
variance of measured       0.368378
mean of differences        0.004170
dtype: float64

## 19. Discussion of the relevance of the Correlation

When comparing the set of measured radii to the set of calculated (predicted) radii one expects a positively linear relationship (assuming the model is good enough). The strength of such relationship is expressed by the corellation. After the simplification steps (isotropy) discussed above a 0.338 correlation can be regarded as good.

## 20. Discussion Paragraph for the Manuskript

How good is the model?

What does the model say?

For what other problems would the model be applicable?